UART Console: Add option to suppress echo of input

dwld opened the pull request:
When using UART for "asynchronous" communication,
the echo of the input may break the output to read.

With the introduction of a new config option
CONFIG_UART_CONSOLE_SUPPRESS_ECHO,
the echo may be suppressed if needed.

Signed-off-by: Stefan Diewald <dwld@users.noreply.github.com>

dwld committed:
UART Console: Add option to suppress echo of input

When using UART for "asynchronous" communication,
the echo of the input may break the output to read.

With the introduction of a new config option
CONFIG_UART_CONSOLE_SURPRESS_ECHO,
the echo may be suppressed if needed.

Signed-off-by: Stefan Diewald <dwld@users.noreply.github.com>

dwld committed:
UART Console: Fixed checkpath issue on whitespace in Kconfig

Fixed checkpath issue on whitespace in Kconfig

Signed-off-by: Stefan Diewald <dwld@users.noreply.github.com>

jfischer-phytec-iot commented on drivers/console/Kconfig:
```diff
@@ -43,6 +43,14 @@ config CONSOLE_HANDLER
 	  This option enables console input handler allowing to write simple
 	  interaction between serial console and the OS.
+	  
+config UART_CONSOLE_SUPPRESS_ECHO
```
Should be below `config UART_CONSOLE` block

jfischer-phytec-iot commented on drivers/console/uart_console.c:
```diff
@@ -170,8 +170,10 @@ static void insert_char(char *pos, char c, u8_t end)
 {
 	char tmp;
 
+#ifndef CONFIG_UART_CONSOLE_SUPPRESS_ECHO
```
If necessary at all, would it not be a prettier solution ?

```diff
diff --git a/drivers/console/uart_console.c b/drivers/console/uart_console.c
index f7d2a8099d..37a581cc69 100644
--- a/drivers/console/uart_console.c
+++ b/drivers/console/uart_console.c
@@ -30,6 +30,7 @@
 #include <linker/sections.h>
 #include <sys/atomic.h>
 #include <sys/printk.h>
+#include <sys/util.h>
 #ifdef CONFIG_UART_CONSOLE_MCUMGR
 #include "mgmt/serial.h"
 #endif
@@ -166,12 +167,19 @@ static inline void cursor_restore(void)
        printk("\x1b[u");
 }
 
+static void console_echo_back(struct device *dev, char c)
+{
+       if (!(IS_ENABLED(CONFIG_UART_CONSOLE_SUPPRESS_ECHO))) {
+               uart_poll_out(dev, c);
+       }
+}
+
 static void insert_char(char *pos, char c, u8_t end)
 {
        char tmp;
 
        /* Echo back to console */
-       uart_poll_out(uart_console_dev, c);
+       console_echo_back(uart_console_dev, c);
 
        if (end == 0U) {
                *pos = c;
@@ -184,7 +192,7 @@ static void insert_char(char *pos, char c, u8_t end)
        cursor_save();
 
        while (end-- > 0) {
-               uart_poll_out(uart_console_dev, tmp);
+               console_echo_back(uart_console_dev, tmp);
                c = *pos;
                *(pos++) = tmp;
                tmp = c;
@@ -196,11 +204,11 @@ static void insert_char(char *pos, char c, u8_t end)
 
 static void del_char(char *pos, u8_t end)
 {
-       uart_poll_out(uart_console_dev, '\b');
+       console_echo_back(uart_console_dev, '\b');
 
        if (end == 0U) {
-               uart_poll_out(uart_console_dev, ' ');
-               uart_poll_out(uart_console_dev, '\b');
+               console_echo_back(uart_console_dev, ' ');
+               console_echo_back(uart_console_dev, '\b');
                return;
        }
 
@@ -208,10 +216,10 @@ static void del_char(char *pos, u8_t end)

        while (end-- > 0) {
                *pos = *(pos + 1);
-               uart_poll_out(uart_console_dev, *(pos++));
+               console_echo_back(uart_console_dev, *(pos++));
        }
 
-       uart_poll_out(uart_console_dev, ' ');
+       console_echo_back(uart_console_dev, ' ');
 
        /* Move cursor back to right place */
        cursor_restore();
@@ -508,8 +516,8 @@ void uart_console_isr(struct device *unused)
                                break;
                        case '\r':
                                cmd->line[cur + end] = '\0';
-                               uart_poll_out(uart_console_dev, '\r');
-                               uart_poll_out(uart_console_dev, '\n');
+                               console_echo_back(uart_console_dev, '\r');
+                               console_echo_back(uart_console_dev, '\n');
                                cur = 0U;
                                end = 0U;
                                k_fifo_put(lines_queue, cmd);
```

dwld replied:
Thanks. Yes, that would be much prettier.

dwld committed:
UART Console: Prettified echo suppression

Changed location of config option in Kconfig;
Prettier solution of suppressing the echo as suggested by
@jfischer-phytec-iot

Signed-off-by: Stefan Diewald <dwld@users.noreply.github.com>
